### PR TITLE
Add strong typing for the following API methods

### PR DESCRIPTION
**Pull request by jesusbv**
* kiwi/system/kernel.py
* kiwi/system/result.py
* kiwi/system/shell.py

This references issue #1644

### CODE REVIEW

File: kiwi/system/kernel.py [hunk 1]
```diff
@@ -15,15 +15,23 @@
...
 from kiwi.command import Command
 
 from kiwi.exceptions import KiwiKernelLookupError
 
+kernel_tuple = NamedTuple('kernel', [('name', str),
+                                     ('filename', str),
+                                     ('version', str)])
+xen_hyperv = NamedTuple('xen', [('filename', str),
+                                ('name', str)])
+
```
Comment (schaefi) [inline kiwi/system/kernel.py:34]: Regarding the use of a tuple as type there is a hidden policy in kiwi to name them XX_type and use the same indentation style as everywhere else. Therefore can be above be written as

```python
kernel_type = NamedTuple(
    'kernel_type', [
        ('name', str),
        ('filename', str),
        ('version', str)
    ]
)

xen_hypervisor_type = NamedTuple(
    'xen_hypervisor_type', [
        ('filename', str),
        ('name', str)
    ]
)
```

File: kiwi/system/kernel.py [hunk 2]
```diff
@@ -109,14 +114,16 @@ def copy_kernel(self, target_dir, file_name=None):
         """
         kernel = self.get_kernel()
         if kernel:
-            if not file_name:
+            if file_name is None:
```
Comment (schaefi) [inline kiwi/system/kernel.py:117]: This is a behavior change. ```not str``` matches empty strings and None. ```str is None``` only matches if the string is of the None type. An empty string ```str=""``` would not be matched. Any particular reason for this change ?

File: kiwi/system/kernel.py [hunk 3]
```diff
@@ -128,7 +135,7 @@ def copy_xen_hypervisor(self, target_dir, file_name=None):
         """
         xen = self.get_xen_hypervisor()
         if xen:
-            if not file_name:
+            if file_name is None:
```
Comment (schaefi) [inline kiwi/system/kernel.py:138]: same here, behavior change

File: kiwi/system/result.py [hunk 4]
```diff
@@ -15,45 +15,53 @@
...
 from kiwi.exceptions import (
     KiwiResultError
 )
 
 log = logging.getLogger('kiwi')
 
 # must be global to allow pickle to find it
-result_file_type = namedtuple(
-    'result_file_type', ['filename', 'use_for_bundle', 'compress', 'shasum']
+ResultFileType = NamedTuple(
```
Comment (schaefi) [inline kiwi/system/result.py:34]: Please keep the naming as ```result_file_type``` and the indentation as suggested in former comments

File: kiwi/system/result.py [hunk 5]
```diff
@@ -15,45 +15,53 @@
...
+    'ResultFileType', [('filename', str),
+                       ('use_for_bundle', bool),
+                       ('compress', bool),
+                       ('shasum', bool)]
 )
+ResultType = TypeVar('ResultType', bound='Result')
```
Comment (schaefi) [inline kiwi/system/result.py:40]: I did not understand why we need a new global ResultType variable from the Result class. Is a self reference to the Result class not possible when used as return type in ```def load(filename: str) -> ResultType:``` ?

**Review by schaefi — CHANGES_REQUESTED** [review]
Overall looks good, thanks :+1: I was concerned about one behavior change as I think and having one question regarding the ResultType setting

Thoughts ?

File: kiwi/system/result.py (hunk 5 above)
Comment (schaefi) [reply kiwi/system/result.py:40]: meaning ```def load(filename: str) -> Result:``` does not work ?

File: kiwi/system/kernel.py (hunk 2 above)
Comment (jesusbv) [reply kiwi/system/kernel.py:117]: I missed the empty string check but the reason for the change is that when deciding the changes for strong typing I got an
error asking for explicit check in those lines. I will revert it as it is wrong and not explicit check needed anymore

File: kiwi/system/result.py (hunk 4 above)
Comment (jesusbv) [reply kiwi/system/result.py:34]: I will change it.

The reason for not choosing snake case naming was that I get an info message from mypy
```
 Class name "result_file_type" doesn't conform to PascalCase naming style
 ```

File: kiwi/system/kernel.py (hunk 1 above)
Comment (jesusbv) [reply kiwi/system/kernel.py]: Done.

**Review by schaefi — APPROVED** [review]
Looks great :+1:

**Pull request merged**